### DISCUSSION TITLE
Super class combining Image with ButtonBehaviour does not work as a button when setting size_hint from formula (Python)

### DISCUSSION THREAD

**Discussion by FilipKosel** [kivy/kivy-github-discussions#43] [Q&A]
Hello all,

First post on here, apologies in advance for any issues. I'll try to keep this short.

TL;DR: Super class combining ButtonBehaviour with Image does not work as a button when size_hint is set using formula to calculate values.

We're using Kivy in a Python project to create touchscreen tasks (see [https://github.com/FilipKosel/TouchCog](https://github.com/FilipKosel/TouchCog) for more details). For reference, all Kivy coding is done directly in Python (i.e., we do not use .kv files); this is how the project was started before I joined on and this will not change in the foreseeable future.

We have an ImageButton super class that combines ButtonBehaviour with Image as follows:

```
class ImageButton(ButtonBehavior, Image):

	def __init__(self, **kwargs):

		super(ImageButton, self).__init__(**kwargs)

		self.coord = None
		self.fit_mode = 'fill'
		self.touch_pos = (0, 0)
		self.name = ''
```

In most situations, we hardcode the size_hint for each ImageButton, and this works well; the images show up as the appropriate size and function as buttons if the user touches within the boundaries of the image. However, in one scenario, the screen is divided into a grid of equally-sized squares, although the number of squares for each axis can change. Please see the following (simplified) code block:

```
grid_square_x_size = (grid_x_pos_list[-1] - grid_x_pos_list[0]) / (len(grid_x_pos_list) - 1)

grid_square_y_size = (grid_y_pos_list[-1] - grid_y_pos_list[0]) / (len(grid_y_pos_list) - 1)

grid_target_button.size_hint = (grid_square_x_size, grid_square_y_size)

grid_target_button.bind(on_press=target_button_pressed)
```

For the above:

- grid_square_x_size: x dimension of the image (calculated from the number of grid squares on the x-axis)
- grid_square_y_size: y dimension of the image (calculated from the number of grid squares on the y-axis)
- grid_x_pos_list: list containing all possible coordinates for center_x (list is defined elsewhere; changes over time)
- grid_y_pos_list: list containing all possible coordinates for center_y (list is defined elsewhere; changes over time)
- grid_target_button: ImageButton
- target_button_pressed: function called on ImageButton press

This code will display an appropriately-sized image, but the ButtonBehaviour does not work. Based on our troubleshooting:

- Our program registers and records all touch coordinates.
- The code is not hanging anywhere; it continues to execute and will end a given time after start (as intended).
- grid_square_x_size and grid_square_y_size are set to the correct type (float) and values.
- Combining grid_square_x_size and grid_square_y_size into another variable as a tuple, then assigning that variable to size_hint, does not work.
- Setting size_hint_x and size_hint_y separately does not work.
- Using float(), float(str()), or returning values from a function does not work.
- Hard-coding one _or_ both values restores button functionality, even in combination with the above methods (function return not tested).

The only thing I can think of is that the size_hint is correctly passed to the Image component, but not ButtonBehaviour, so the program is not correctly identifying the button boundaries (and therefore is not registering that a touch is occuring within the boundaries).

Does anyone have any insight on how to address this? We've been stuck on this for a couple weeks and I'm running out of ideas.

Thanks in advance!

**Comment by akshayaurora** [comment]
This is not an issue tracker. Please use this forum only for discussions related to the framework , not for issues related to particular repository
